File: backend/app/services/user_service.py
"""
User service for managing users with database
"""

from typing import Optional
from sqlalchemy.orm import Session
from sqlalchemy import select
from app.models.db_models import User
import uuid
import bcrypt


# Password hashing utilities
def get_password_hash(password: str) -> str:
    """Hash a password using bcrypt."""
    if not password:
        raise ValueError("Password cannot be empty")
    
    # Bcrypt has a 72-byte limit, truncate if necessary
    password_bytes = password.encode('utf-8')
    if len(password_bytes) > 72:
        password_bytes = password_bytes[:72]
    
    # Generate salt and hash
    salt = bcrypt.gensalt()
    hashed = bcrypt.hashpw(password_bytes, salt)
    
    # Return as string
    return hashed.decode('utf-8')


def verify_password(plain_password: str, hashed_password: str) -> bool:
    """Verify a password against its hash."""
    if not plain_password or not hashed_password:
        return False
    
    # Bcrypt has a 72-byte limit, truncate if necessary
    password_bytes = plain_password.encode('utf-8')
    if len(password_bytes) > 72:
        password_bytes = password_bytes[:72]
    
    # Ensure hashed_password is bytes
    if isinstance(hashed_password, str):
        hashed_password_bytes = hashed_password.encode('utf-8')
    else:
        hashed_password_bytes = hashed_password
    
    try:
        return bcrypt.checkpw(password_bytes, hashed_password_bytes)
    except Exception:
        return False


def create_user(
    db: Session,
    email: str,
    password: str,
    full_name: str,
    **kwargs
) -> User:
    """Create a new user."""
    # Check if user already exists
    existing_user = db.execute(
        select(User).where(User.email == email.lower())
    ).scalar_one_or_none()
    
    if existing_user:
        raise ValueError("Email already registered")
    
    password_hash = get_password_hash(password)
    user = User(
        id=uuid.uuid4(),
        email=email.lower(),
        password_hash=password_hash,
        full_name=full_name,
        phone=kwargs.get("phone"),
        date_of_birth=kwargs.get("date_of_birth"),
        gender=kwargs.get("gender"),
        skin_type=kwargs.get("skin_type", "normal"),
        preferences=kwargs.get("preferences", {})
    )
    
    db.add(user)
    db.commit()
    db.refresh(user)
    
    return user


def get_user_by_email(db: Session, email: str) -> Optional[User]:
    """Get user by email."""
    return db.execute(
        select(User).where(User.email == email.lower())
    ).scalar_one_or_none()


def get_user_by_id(db: Session, user_id: str) -> Optional[User]:
    """Get user by ID."""
    try:
        user_uuid = uuid.UUID(user_id)
        return db.execute(
            select(User).where(User.id == user_uuid)
        ).scalar_one_or_none()
    except ValueError:
        return None


def authenticate_user(db: Session, email: str, password: str) -> Optional[User]:
    """Authenticate a user with email and password."""
    user = get_user_by_email(db, email)
    if not user:
        return None
    
    if not verify_password(password, user.password_hash):
        return None
    
    return user


def update_user(db: Session, user_id: str, **kwargs) -> Optional[User]:
    """Update user information."""
    user = get_user_by_id(db, user_id)
    if not user:
        return None
    
    for key, value in kwargs.items():
        if value is not None and hasattr(user, key):
            setattr(user, key, value)
    
    db.commit()
    db.refresh(user)
    
    return user

File: backend-node/src/services/userService.ts
import { AppDataSource } from '../database/connection';
import { User } from '../models/User';
import { hashPassword, verifyPassword } from '../utils/password';

/**
 * Create a new user
 */
export async function createUser(
  email: string,
  password: string,
  fullName: string,
  phone?: string,
  dateOfBirth?: string,
  gender?: string,
  skinType: string = 'normal',
  preferences: Record<string, any> = {}
): Promise<User> {
  const userRepository = AppDataSource.getRepository(User);

  // Check if user already exists
  const existingUser = await userRepository.findOne({
    where: { email: email.toLowerCase() },
  });

  if (existingUser) {
    throw new Error('Email already registered');
  }

  // Hash password
  const passwordHash = await hashPassword(password);

  // Create user
  const user = userRepository.create({
    email: email.toLowerCase(),
    passwordHash,
    fullName,
    phone: phone || null,
    dateOfBirth: dateOfBirth || null,
    gender: gender || null,
    skinType,
    preferences,
  });

  await userRepository.save(user);
  return user;
}

/**
 * Get user by email
 */
export async function getUserByEmail(email: string): Promise<User | null> {
  const userRepository = AppDataSource.getRepository(User);
  return userRepository.findOne({
    where: { email: email.toLowerCase() },
  });
}

/**
 * Get user by ID
 */
export async function getUserById(userId: string): Promise<User | null> {
  const userRepository = AppDataSource.getRepository(User);
  return userRepository.findOne({
    where: { id: userId },
  });
}

/**
 * Authenticate a user with email and password
 */
export async function authenticateUser(
  email: string,
  password: string
): Promise<User | null> {
  const user = await getUserByEmail(email);
  if (!user) {
    return null;
  }

  const isValid = await verifyPassword(password, user.passwordHash);
  if (!isValid) {
    return null;
  }

  return user;
}

/**
 * Update user information
 */
export async function updateUser(
  userId: string,
  updates: {
    fullName?: string;
    phone?: string;
    dateOfBirth?: string;
    gender?: string;
    skinType?: string;
    preferences?: Record<string, any>;
  }
): Promise<User | null> {
  const userRepository = AppDataSource.getRepository(User);
  const user = await getUserById(userId);

  if (!user) {
    return null;
  }

  // Update fields
  if (updates.fullName !== undefined) user.fullName = updates.fullName;
  if (updates.phone !== undefined) user.phone = updates.phone;
  if (updates.dateOfBirth !== undefined) user.dateOfBirth = updates.dateOfBirth;
  if (updates.gender !== undefined) user.gender = updates.gender;
  if (updates.skinType !== undefined) user.skinType = updates.skinType;
  if (updates.preferences !== undefined) user.preferences = updates.preferences;

  await userRepository.save(user);
  return user;
}

